lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+  rewriter.setInsertionPointToStart(op.getBody());
+
+  SmallVector<ArrayRef<int64_t>> originalShapes{mShape, nShape, kShape};
+  SmallVector<SmallVector<SmallVector<unsigned>>> tilingVectors{
+      {tileShapeM}, {tileShapeN}, {tileShapeM, tileShapeN}};
+
+  // Replace old args with newly computed args
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++) {
+    int operandIndex = 0;
+    for (auto arg : oper->getOperands()) {
+      int oldArgIndex = -1;
+      for (int i = 0; i < numArgs; i++) {
+        if (arg == op.getInductionVar(i)) {
+          oldArgIndex = i;
+          break;
+        }
+      }
+      if (oldArgIndex != -1) {
+        Value add = op.getBody()->getArgument(tileOffsets[oldArgIndex]);
+        Value mul;
+        for (int j = tileOffsets[oldArgIndex] + 1;
+             j < tileOffsets[oldArgIndex + 1]; j++) {
+          unsigned cumulativeUpperBound = 1;
+          for (int k = j; k < tileOffsets[oldArgIndex + 1]; k++)
+            cumulativeUpperBound *= op.getStaticUpperBound()[j];
+          Value upperBound = rewriter.create<arith::ConstantIndexOp>(
+              op.getLoc(), cumulativeUpperBound);
+          mul = rewriter.create<arith::MulIOp>(op.getLoc(), b.getIndexType(),
+                                               add, upperBound);
+          add = rewriter.create<arith::AddIOp>(
+              op.getLoc(), b.getIndexType(), mul, op.getBody()->getArgument(j));
+        }
+        oper->setOperand(operandIndex, add);
+      }
+      operandIndex++;
+    }
+  }
+
+  for (int i = 1; i <= 3; i++) {
+    auto definingOp = brgemmOp.getOperand(i).getDefiningOp();
+    if (isa<memref::SubViewOp>(definingOp)) {
+      Value operand = definingOp->getOperand(0);
+
+      auto operandType = operand.getType();
+      auto resultType = dyn_cast<MemRefType>(brgemmOp.getOperand(i).getType());
+      auto reassociationIndex =
+          getReassociationIndices(originalShapes[i - 1], tilingVectors[i - 1]);
+
+      SmallVector<int64_t> shape;
+      for (size_t j = 0; j < tilingVectors[i - 1].size(); j++)
+        shape.append(tilingVectors[i - 1][j].begin(),
+                     tilingVectors[i - 1][j].end());
+      shape.append(
+          std::next(originalShapes[i - 1].begin(), tilingVectors[i - 1].size()),
+          originalShapes[i - 1].end());
+      int inductionVarBase = i - 1;
+      if (definingOp == brgemmOp.getOperand(3).getDefiningOp())
+        inductionVarBase = 0;
+      insertSubview(shape, operandType, resultType, reassociationIndex, operand,
+                    op, rewriter, definingOp, inductionVarBase);
+    }
+  }
+
+  return success();
+}
+
+bool getInnermostForLoops(Operation *rootOp,
+                          SmallVectorImpl<scf::ForallOp> &result) {
+  assert(rootOp != nullptr && "Root operation must not be a nullptr.");
+  bool rootEnclosesForAllloops = false;
+  for (Region &region : rootOp->getRegions()) {
```

Comment:
Isn't there a better way to walk the ops of an enclosing op?

Reply:
We can use walk but that will also translate to a recursive walk eventually, because we look for the innermost scf forall loop.